provisioning_client/adapters/hsm_client_x509.c

```diff
@@ -0,0 +1,196 @@
+#include "azure_c_shared_utility/urlencode.h"
+#include "azure_c_shared_utility/xlogging.h"
+#include "azure_c_shared_utility/crt_abstractions.h"
+
+#include "hsm_client_x509.h"
+#include "hsm_client_data.h"
+
+typedef struct HSM_CLIENT_X509_INFO_TAG
+{
+    const char* x509certificate;
+    const char* x509key;
+} HSM_CLIENT_X509_INFO;
+
+static HSM_CLIENT_X509_INFO hsm_singleton;
+
+HSM_CLIENT_HANDLE hsm_client_x509_create(void)
+{
+    HSM_CLIENT_X509_INFO* result;
+    result = &hsm_singleton;
+    return result;
+}
+
+void hsm_client_x509_destroy(HSM_CLIENT_HANDLE handle)
+{
+    (void)handle;
+    // No-op.
+}
+
+int hsm_client_x509_set_certificate(HSM_CLIENT_HANDLE handle, const char* certificate)
+{
+    int result;
+    if (handle == NULL || certificate == NULL)
+    {
+        LogError("Invalid parameter handle: %p, certificate: %p", handle, certificate);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        HSM_CLIENT_X509_INFO* x509_client = (HSM_CLIENT_X509_INFO*)handle;
+        if (x509_client->x509certificate != NULL)
+        {
+            LogError("Certificate has been previously set and cannot be changed");
+            result = MU_FAILURE;
+        }
+        else if (mallocAndStrcpy_s((char **)&(x509_client->x509certificate), certificate) != 0)
+        {
+            LogError("Failed allocating certificate key");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            result = 0;
+        }
+    }
+    return result;
+}
+
+int hsm_client_x509_set_key(HSM_CLIENT_HANDLE handle, const char* key)
+{
+    int result;
+    if (handle == NULL || key == NULL)
+    {
+        LogError("Invalid parameter handle: %p, key: %p", handle, key);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        HSM_CLIENT_X509_INFO* x509_client = (HSM_CLIENT_X509_INFO*)handle;
+        if (x509_client->x509key != NULL)
+        {
+            LogError("Certificate key has been previously set and cannot be changed");
+            result = MU_FAILURE;
+        }
+        else if (mallocAndStrcpy_s((char **)&x509_client->x509key, key) != 0)
+        {
+            LogError("Failed allocating cert key");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            result = 0;
+        }
+    }
+    return result;
+}
+
+char* hsm_client_x509_get_certificate(HSM_CLIENT_HANDLE handle)
+{
+    char* result;
+    if (handle == NULL)
+    {
+        LogError("Invalid client handle value specified.");
+        result = NULL;
+    }
+    else
+    {
+        HSM_CLIENT_X509_INFO* x509_client = (HSM_CLIENT_X509_INFO*)handle;
+
+        if (x509_client->x509certificate == NULL)
+        {
+            LogError("Client X509 certificate was not configured.");
+            result = NULL;
+        }
+        else
+        {
+            // Create a copy. Lifetime is managed by the DPS or Hub clients.
+            if (mallocAndStrcpy_s(&result, x509_client->x509certificate) != 0)
```

Comment:
curious why this can't just pass back the pointer to the stored cert? it would be valid until the user calls `void hsm_client_x509_deinit(void)`?

Reply:
same goes for the key

Reply:
I agree and have it that way originally. 
It's because of semantics established within IoT Hub client for these APIs / set options: We always create copies and allow the app to deallocate.